src/gc.c

```diff
@@ -1953,15 +2005,20 @@ mark: {
 #endif
         jl_taggedvalue_t *o = jl_astaggedvalue(new_obj);
         jl_datatype_t *vt = (jl_datatype_t*)tag;
+        int foreign_alloc = ((void*)o >= sysimg_base && (void*)o < sysimg_end);
         int update_meta = __likely(!meta_updated && !gc_verifying);
         meta_updated = 0;
         // Symbols are always marked
         assert(vt != jl_symbol_type);
         if (vt == jl_simplevector_type) {
             size_t l = jl_svec_len(new_obj);
             jl_value_t **data = jl_svec_data(new_obj);
-            if (update_meta)
-                gc_setmark(ptls, o, bits, l * sizeof(void*) + sizeof(jl_svec_t));
+            if (update_meta) {
+                if (foreign_alloc)
```

Comment:
Since `objprofile_count` is a non-default debug option, I think it'll be better to make `update_meta = 0` in this case and do this in an else branch so that it's more clear to the compiler that the two branches can be merged together. It also seems like it's exactly the same code so maybe just do at the very beginning

```c
if (update_meta) {
    if (((void*)o >= sysimg_base && (void*)o < sysimg_end)) {
        update_meta = 0;
        objprofile_count(vt, bits == GC_OLD_MARKED, jl_datasize_size(vt));
    }
}
```